src/public/modules/forms/helpers/response-factory.ts

```diff
@@ -0,0 +1,77 @@
+import { hasProp } from '../../../../shared/util/has-prop'
+import {
+  ArrayResponse,
+  DisplayedResponseWithoutAnswer,
+  NestedResponse,
+  SingleResponse,
+} from '../../../../types/response'
+
+import {
+  ArrayAnswerResponse,
+  Response,
+  SingleAnswerResponse,
+  TableResponse,
+} from './csv-response-classes'
+
+/**
+ * Converts a field record into a custom response instance
+ * @param fieldRecordData Field record
+ * @returns Response instance
+ * @throws Error when data does not fit any known response type
+ */
+export const getResponseInstance = (
+  fieldRecordData: DisplayedResponseWithoutAnswer,
+): Response => {
+  if (
+    isNestedResponse(fieldRecordData) &&
+    fieldRecordData.fieldType === 'table'
+  ) {
+    return new TableResponse(fieldRecordData)
+  } else if (
+    isArrayResponse(fieldRecordData) &&
+    fieldRecordData.fieldType === 'checkbox'
+  ) {
+    return new ArrayAnswerResponse(fieldRecordData)
+  } else if (isSingleResponse(fieldRecordData)) {
+    return new SingleAnswerResponse(fieldRecordData)
+  } else {
+    // eslint-disable-next-line typesafe/no-throw-sync-func
+    throw new Error('Response did not match any known type')
+  }
+}
+
+const isNestedResponse = (
+  response: DisplayedResponseWithoutAnswer,
+): response is NestedResponse => {
+  return (
+    hasAnswerArray(response) &&
+    response.answerArray.every(
+      (arr) =>
+        Array.isArray(arr) &&
+        arr.every((value: unknown) => typeof value === 'string'),
+    )
+  )
+}
+
+const isArrayResponse = (
+  response: DisplayedResponseWithoutAnswer,
+): response is ArrayResponse => {
+  return (
+    hasAnswerArray(response) &&
+    response.answerArray.every((value) => typeof value === 'string')
+  )
+}
+
+const isSingleResponse = (
+  response: DisplayedResponseWithoutAnswer,
+): response is SingleResponse => {
+  return hasProp(response, 'answer') && typeof response.answer === 'string'
```

Comment:
i think there may be issues with this check; are we certain that all answers must be string? What about number fields? They might be of `number` type instead?

Reply:
Responses are decrypted via the `decrypt` function in the crypto SDK which returns a `DecryptedObject` which has properties `responses: FormField[]` and `verified?: Record<string, any>`. Answers in `FormField` are supposed to be of type 
```
{ answer: string; answerArray?: never }
  | { answer?: never; answerArray: string[] | string[][] }
```
`responses` and `verified` are then combined after processing the verified fields and form the csv records. So I think number field answers are also strings after decryption. I'm not sure if we should be removing the check `typeof response.answer === 'string'` because verified field answers could be of type `any`. But I think it still works because if I'm not wrong, all verified field answers are also strings at the moment.

Could you confirm if my understanding is correct and whether you think we should remove the check I mentioned?

Reply:
Should be fine then, make sure to test on all form field types just in case :P

Reply:
I think the e2e tests do check the downloaded csv content for the case when all fields are added. That should be enough right?

Reply:
yea, don't think there's a way to programatically test them correctly since we will be mocking everything anyways